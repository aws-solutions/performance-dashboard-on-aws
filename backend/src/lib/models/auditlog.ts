/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum ItemEvent {
    Create = "Create",
    Update = "Update",
    Delete = "Delete",
}

export type ModifiedProperty = {
    property: string;
    oldValue: any;
    newValue: any;
};

export interface AuditLog {
    itemId: string;
    timestamp: Date;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

// Represents a dynamodb item in the Audit Trail table
export interface AuditLogItem {
    pk: string;
    sk: string;
    type: string;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

export interface DashboardAuditLog extends AuditLog {
    version: number;
}

// Represents a dynamodb item in the Audit Trail table that
// contains attributes specific to Dashboard audit logs.
export interface DashboardAuditLogItem extends AuditLogItem {
    version: number;
}
