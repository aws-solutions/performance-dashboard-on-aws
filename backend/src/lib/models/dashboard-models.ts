export type Dashboard = {
    id: string,
    name: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string,
    sk: string,
    type: string,
    dashboardName: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};