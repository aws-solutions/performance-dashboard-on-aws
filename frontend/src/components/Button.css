.usa-button--base {
  background-color: #565c65 !important;
}
.usa-button--base:hover,
.usa-button--base.usa-button--hover {
  background-color: #3d4551 !important;
}
.usa-button--base:active,
.usa-button--base.usa-button--active {
  background-color: #1b1b1b !important;
}
.usa-button--base:disabled {
  background-color: #c9c9c9 !important;
}
