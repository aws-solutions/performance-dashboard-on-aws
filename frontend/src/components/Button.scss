@import "../styles/base";

.usa-button--outline {
    @include u-text("base", "!important");
    background-color: transparent !important;
    box-shadow: inset 0 0 0 2px color("base") !important;
}
.usa-button--outline:visited {
    @include u-text("base", "!important");
}
.usa-button--outline:hover,
.usa-button--outline.usa-button--hover {
    @include u-text("base-dark", "!important");
    background-color: transparent !important;
    box-shadow: inset 0 0 0 2px color("base-dark") !important;
}
.usa-button--outline:active,
.usa-button--outline.usa-button--active {
    @include u-text("base-dark", "!important");
    background-color: transparent !important;
    box-shadow: inset 0 0 0 2px color("base-dark") !important;
}

.usa-button--outline:disabled,
.usa-button--outline.usa-button--disabled {
    @include u-text("base-light", "!important");
    background-color: transparent !important;
    box-shadow: inset 0 0 0 2px color("base-light") !important;
}
