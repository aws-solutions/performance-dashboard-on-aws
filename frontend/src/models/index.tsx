/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export type TopicArea = {
    id: string;
    name: string;
    createdBy: string;
    dashboardCount: number;
};

export const TopicAreaSortingCriteria = (a: TopicArea, b: TopicArea) => (a.name > b.name ? 1 : -1);

export type PublicTopicArea = {
    id: string;
    name: string;
    dashboards?: Array<Dashboard | PublicDashboard>;
};

export enum DashboardState {
    Draft = "Draft",
    Published = "Published",
    Archived = "Archived",
    PublishPending = "PublishPending",
    Inactive = "Inactive",
}

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetType {
    DynamicDataset = "DynamicDataset",
    StaticDataset = "StaticDataset",
    CsvFileUpload = "CsvFileUpload",
    CreateNew = "CreateNew",
}

export type Dashboard = {
    id: string;
    name: string;
    version: number;
    parentDashboardId: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    releaseNotes?: string;
    widgets: Array<Widget>;
    state: string;
    updatedAt: Date;
    createdBy: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type PublicDashboard = {
    id: string;
    name: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    widgets: Array<Widget>;
    updatedAt: Date;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type DashboardVersion = {
    id: string;
    version: number;
    state: DashboardState;
    friendlyURL?: string;
};

export type DashboardAuditLog = {
    itemId: string;
    timestamp: Date;
    version: number;
    userId: string;
    event: "Create" | "Update" | "Delete";
    modifiedProperties?: Array<{
        property: string;
        oldValue: any;
        newValue: any;
    }>;
};

export enum WidgetType {
    Text = "Text",
    Chart = "Chart",
    Table = "Table",
    Image = "Image",
    Metrics = "Metrics",
    Section = "Section",
}

export enum ChartType {
    LineChart = "LineChart",
    ColumnChart = "ColumnChart",
    BarChart = "BarChart",
    PartWholeChart = "PartWholeChart",
    PieChart = "PieChart",
    DonutChart = "DonutChart",
}

export interface Widget {
    id: string;
    name: string;
    widgetType: string;
    order: number;
    updatedAt: Date;
    dashboardId: string;
    content: any;
    showTitle: boolean;
    section?: string;
}

export interface ChartWidget extends Widget {
    content: {
        title: string;
        chartType: ChartType;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        horizontalScroll?: boolean;
        stackedChart?: boolean;
        significantDigitLabels: boolean;
        dataLabels: boolean;
        computePercentages: boolean;
        showTotal: boolean;
    };
}

export interface TableWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        significantDigitLabels: boolean;
        displayWithPages: boolean;
    };
}

export interface ImageWidget extends Widget {
    content: {
        title: string;
        imageAltText: string;
        summary?: string;
        summaryBelow: boolean;
        s3Key: {
            raw: string;
        };
        filename: string;
        scalePct?: string;
    };
}

export interface MetricsWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        oneMetricPerRow: boolean;
        datasetType?: DatasetType;
        significantDigitLabels: boolean;
        metricsCenterAlign: boolean;
        s3Key: {
            raw: string;
            json: string;
        };
    };
}

export interface SectionWidget extends Widget {
    content: {
        title: string;
        summary: string;
        widgetIds?: Array<string>;
        showWithTabs: boolean;
        horizontally?: boolean;
    };
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export type Dataset = {
    id: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    createdBy?: string;
    updatedAt?: Date;
    sourceType?: SourceType;
    schema?: DatasetSchema;
};

export type PublicHomepage = {
    title: string;
    description: string;
    dashboards: Array<PublicDashboard>;
};

export type Homepage = {
    title: string;
    description: string;
    updatedAt: Date;
};

export type PublicSettings = {
    dateTimeFormat: {
        date: string;
        time: string;
    };
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels?: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Settings = {
    publishingGuidance: string;
    dateTimeFormat: {
        date: string;
        time: string;
    };
    updatedAt?: Date;
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customLogoAltText?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    adminContactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Metric = {
    title: string;
    value: number;
    changeOverTime?: string;
    percentage?: string;
    currency?: string;
    startDate?: string;
    endDate?: string;
};

// Type for the History object in react-router. Defines the
// location state that is common across all routes.
// Usage in a component: useHistory<LocationState>();
export type LocationState = {
    alert?: {
        type: "success" | "warning" | "info" | "error";
        message: string;
        to?: string;
        linkLabel?: string;
    };
    emails?: string;
    usernames?: Array<string>;
    id?: string;
    metrics?: Array<Metric>;
    metric?: Metric;
    position?: number;
    showTitle?: boolean;
    oneMetricPerRow?: boolean;
    significantDigitLabels?: boolean;
    metricsCenterAlign?: boolean;
    metricTitle?: string;
    origin?: string;
    json?: Array<any>;
    staticDataset?: Dataset;
    redirectUrl?: string;
    crumbLabel?: string;
    datasetType?: DatasetType;
};

export enum UserRoles {
    Admin = "Admin",
    Editor = "Editor",
    /**
     * @type {string}
     * @deprecated no longer supported
     */
    Publisher = "Publisher",
    Public = "Public",
}

export type User = {
    userId: string;
    enabled: boolean;
    userStatus: string;
    sub: string;
    email: string;
    roles: Array<UserRoles>;
    createdAt: Date;
    updatedAt: Date;
};

export enum ColumnDataType {
    Number = "Number",
    Text = "Text",
    Date = "Date",
}

export enum NumberDataType {
    Percentage = "Percentage",
    Currency = "Currency",
    "With thousands separators" = "With thousands separators",
}

export enum CurrencyDataType {
    "Dollar $" = "Dollar $",
    "Euro €" = "Euro €",
    "Pound £" = "Pound £",
}

export interface ColumnMetadata {
    columnName: string;
    dataType?: ColumnDataType;
    numberType?: NumberDataType;
    currencyType?: CurrencyDataType;
    hidden: boolean;
}
