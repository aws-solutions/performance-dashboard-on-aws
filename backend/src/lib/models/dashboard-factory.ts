import { v4 as uuidv4 } from "uuid";
import { Dashboard, DashboardItem } from "./dashboard-models";
import { User } from "./user-models";
import topicareaFactory from "./topicarea-factory";

const DASHBOARD: string = 'Dashboard';

function createNew(name: string, topicAreaId: string, topicAreaName: string, user: User): Dashboard {
  return {
    id: uuidv4(),
    name,
    topicAreaId,
    topicAreaName,
    createdBy: user.userId,
  };
}

/**
 * Converts a Dashboard to a DynamoDB item.
 */
function toItem(dashboard: Dashboard): DashboardItem {
  let item: DashboardItem = {
    pk: itemId(dashboard.id),
    sk: itemId(dashboard.id),
    type: DASHBOARD,
    dashboardName: dashboard.name,
    topicAreaName: dashboard.topicAreaName,
    topicAreaId: topicareaFactory.itemId(dashboard.topicAreaId),
    description: dashboard.description,
    overview: dashboard.overview,
    createdBy: dashboard.createdBy,
  };
  return item;
}

/**
 * Converts a DynamoDB item into a Dashboard object
 */
function fromItem(item: DashboardItem): Dashboard {
  const id = item.pk.substring(10);
  let dashboard: Dashboard = {
    id,
    name: item.dashboardName,
    topicAreaId: item.topicAreaId.substring(10),
    topicAreaName: item.topicAreaName,
    description: item.description,
    overview: item.overview,
    createdBy: item.createdBy,
  }
  return dashboard;
}

function itemId(id: string): string { return `${DASHBOARD}#${id}` }

export default {
  toItem,
  createNew,
  fromItem,
  itemId,
};
