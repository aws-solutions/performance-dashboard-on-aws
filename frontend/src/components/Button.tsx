/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React, { KeyboardEvent, MouseEvent, ReactNode } from "react";
import "./Button.scss";

type Variant = "base" | "default" | "outline" | "secondary" | "accent" | "unstyled";

interface Props {
    children: ReactNode;
    onClick?: (event: MouseEvent<HTMLButtonElement>) => void;
    onKeyDown?: (event: KeyboardEvent<HTMLButtonElement>) => void;
    variant?: Variant;
    className?: string;
    testid?: string;
    disabled?: boolean;
    type?: "submit" | "reset" | "button";
    ariaLabel?: string;
    ariaCurrent?: boolean | "time" | "false" | "true" | "page" | "step" | "location" | "date";
    ariaExpanded?: boolean;
    ariaControls?: string;
    disabledToolTip?: string;
}

const Button = React.forwardRef<HTMLButtonElement, Props>((props, ref) => {
    let variantClassName = "";
    switch (props.variant) {
        case "base":
            variantClassName = " usa-button--base";
            break;
        case "outline":
            variantClassName = " usa-button--outline";
            break;
        case "secondary":
            variantClassName = " usa-button--secondary";
            break;
        case "accent":
            variantClassName = " usa-button--accent-cool";
            break;
        case "unstyled":
            variantClassName = " usa-button--unstyled";
            break;
        default:
            variantClassName = " usa-button--base";
            break;
    }

    let additionalClasses = "";
    if (props.className) {
        additionalClasses = ` ${props.className}`;
    }

    const button = (
        <button
            data-testid={props.testid}
            aria-label={props.ariaLabel}
            aria-current={props.ariaCurrent}
            aria-expanded={props.ariaExpanded}
            aria-controls={props.ariaControls}
            className={`usa-button${variantClassName}${additionalClasses}`}
            disabled={props.disabled}
            ref={ref}
            type={props.type}
            onClick={(e) => {
                if (props.onClick) {
                    props.onClick(e);
                }
            }}
            onKeyDown={(e) => {
                if (props.onKeyDown) {
                    props.onKeyDown(e);
                }
            }}
        >
            {props.children}
        </button>
    );

    // show disabled tooltips
    if (props.disabled && props.disabledToolTip) {
        return (
            <span data-toggle="tooltip" data-placement="top" title={props.disabledToolTip}>
                {button}
            </span>
        );
    }

    return button;
});

export default Button;
