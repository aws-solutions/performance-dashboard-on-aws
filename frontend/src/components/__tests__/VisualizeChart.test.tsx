/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render, screen } from "@testing-library/react";
import VisualizeChart from "../VisualizeChart";
import { MemoryRouter } from "react-router-dom";
import { ChartType, DatasetType } from "../../models";
import Button from "../Button";

const dataSerie = [
    {
        id: "1",
        name: "Banana",
        updatedAt: "2021-11-11",
    },
    {
        id: "2",
        name: "Chocolate",
        updatedAt: "2020-11-11",
    },
    {
        id: "3",
        name: "Vanilla",
        updatedAt: "2019-11-11",
    },
];
test("renders the VisualizeChart component", async () => {
    const wrapper = render(
        <VisualizeChart
            errors={[]}
            register={() => {}}
            json={dataSerie}
            csvJson={[]}
            datasetLoading={false}
            datasetType={DatasetType.DynamicDataset}
            onCancel={() => {}}
            backStep={() => {}}
            advanceStep={() => {}}
            fileLoading={false}
            processingWidget={false}
            fullPreview={false}
            fullPreviewButton={<Button>Full preview</Button>}
            submitButtonLabel="Add chart"
            summaryBelow={false}
            setSortByColumn={() => {}}
            setSortByDesc={() => {}}
            horizontalScroll={true}
            significantDigitLabels={false}
            originalJson={dataSerie}
            chartType={ChartType.LineChart}
            headers={["1", "2", "3"]}
            showTitle={true}
            summary={"summary"}
            title={"title"}
            columnsMetadata={[]}
            configHeader={<></>}
            dataLabels={false}
            showTotal={true}
            widgetId="widget-id"
            previewPanelId="panel-id"
            computePercentages={false}
            stackedChart={false}
        />,
        { wrapper: MemoryRouter },
    );
    expect(screen.getByText("title")).toBeInTheDocument();
    expect(screen.getByText("summary")).toBeInTheDocument();
});
