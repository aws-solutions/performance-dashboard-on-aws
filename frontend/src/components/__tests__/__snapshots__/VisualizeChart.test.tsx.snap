// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the VisualizeChart component 1`] = `
<div>
  <div
    class="grid-row width-desktop"
  >
    <div
      class="grid-col-5"
    >
      <div
        class="usa-form-group"
      >
        <label
          class="usa-label text-bold"
          for="title"
        >
          Chart title
        </label>
        <div
          class="usa-hint"
        >
          Give your chart a descriptive title.
        </div>
        <input
          class="usa-input"
          id="title"
          name="title"
          type="text"
          value=""
        />
      </div>
      <div
        class="usa-checkbox"
      >
        <input
          checked=""
          class="usa-checkbox__input"
          id="display-title"
          name="showTitle"
          type="checkbox"
        />
        <label
          class="usa-checkbox__label"
          for="display-title"
        >
          Show title on dashboard
        </label>
      </div>
      <div
        class="usa-form-group"
      >
        <fieldset
          class="usa-fieldset"
        >
          <legend
            class="usa-sr-only"
          >
            Chart type
          </legend>
          <span
            class="usa-label text-bold"
          >
            Chart type
          </span>
          <div
            class="usa-hint"
          >
            Choose a chart type.
          </div>
          <div
            class="margin-top-2"
          >
            <div
              class="usa-radio"
            >
              <input
                class="usa-radio__input"
                id="BarChart"
                name="chartType"
                type="radio"
                value="BarChart"
              />
              <label
                class="usa-radio__label"
                for="BarChart"
              >
                Bar
              </label>
            </div>
            <div
              class="usa-radio"
            >
              <input
                class="usa-radio__input"
                id="ColumnChart"
                name="chartType"
                type="radio"
                value="ColumnChart"
              />
              <label
                class="usa-radio__label"
                for="ColumnChart"
              >
                Column
              </label>
            </div>
            <div
              class="usa-radio"
            >
              <input
                checked=""
                class="usa-radio__input"
                id="LineChart"
                name="chartType"
                type="radio"
                value="LineChart"
              />
              <label
                class="usa-radio__label"
                for="LineChart"
              >
                Line
              </label>
            </div>
            <div
              class="usa-radio"
            >
              <input
                class="usa-radio__input"
                id="PartWholeChart"
                name="chartType"
                type="radio"
                value="PartWholeChart"
              />
              <label
                class="usa-radio__label"
                for="PartWholeChart"
              >
                Part-to-whole
              </label>
            </div>
          </div>
        </fieldset>
      </div>
      <div
        class="margin-top-3 grid-col-6"
      >
        <div
          class="usa-form-group"
        >
          <label
            class="usa-label text-bold"
            for="sortData"
          >
            Sort data
          </label>
          <div
            class="usa-hint"
            id="event-date-start-hint"
          />
          <select
            class="usa-select"
            id="sortData"
            name="sortData"
          >
            <option
              selected=""
              value=""
            >
              Select an option
            </option>
            <option
              value="id###asc"
            >
              "id" A-Z
            </option>
            <option
              value="id###desc"
            >
              "id" Z-A
            </option>
            <option
              value="name###asc"
            >
              "name" A-Z
            </option>
            <option
              value="name###desc"
            >
              "name" Z-A
            </option>
            <option
              value="updatedAt###asc"
            >
              "updatedAt" A-Z
            </option>
            <option
              value="updatedAt###desc"
            >
              "updatedAt" Z-A
            </option>
          </select>
        </div>
      </div>
      <div>
        <label
          class="usa-label text-bold"
        >
          Chart Options
        </label>
        <div
          class="usa-hint"
        >
          Format your line chart
        </div>
        <div
          class="usa-checkbox"
        >
          <input
            class="usa-checkbox__input"
            id="significantDigitLabels"
            name="significantDigitLabels"
            type="checkbox"
          />
          <label
            class="usa-checkbox__label"
            for="significantDigitLabels"
          >
            Significant digit labels (123K, 1.2M)
          </label>
        </div>
      </div>
      <div
        hidden=""
      >
        <legend
          class="usa-legend text-bold"
        >
          Line chart options
        </legend>
        <div
          class="usa-checkbox"
        >
          <input
            checked=""
            class="usa-checkbox__input"
            id="horizontalScroll"
            name="horizontalScroll"
            type="checkbox"
          />
          <label
            class="usa-checkbox__label"
            for="horizontalScroll"
          >
            Display with horizontal scroll
          </label>
        </div>
      </div>
      <div
        class="usa-form-group"
      >
        <label
          class="usa-label text-bold"
          for="summary"
        >
          Chart summary - optional
        </label>
        <div
          class="usa-hint"
        >
          Give your chart a summary to explain it in more depth. It can also be read by screen readers to describe the chart for those with visual impairments. This field supports markdown.
           
          <a
            aria-label=""
            class="link"
            href="/admin/markdown"
            target="_blank"
          >
            <span>
              View Markdown Syntax
            </span>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-external-link-alt fa-w-16 fa-xs margin-left-05"
              data-icon="external-link-alt"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                fill="currentColor"
              />
            </svg>
          </a>
        </div>
        <textarea
          class="usa-textarea"
          id="summary"
          name="summary"
          rows="5"
          style="height: auto;"
        />
      </div>
      <div
        class="usa-checkbox"
      >
        <input
          class="usa-checkbox__input"
          id="summary-below"
          name="summaryBelow"
          type="checkbox"
        />
        <label
          class="usa-checkbox__label"
          for="summary-below"
        >
          Show summary below chart
        </label>
      </div>
      <br />
      <br />
      <hr />
      <button
        class="usa-button usa-button--outline"
        type="button"
      >
        Back
      </button>
      <button
        class="usa-button usa-button--base"
        disabled=""
        type="submit"
      >
        Add chart
      </button>
      <button
        class="usa-button usa-button--unstyled text-base-dark hover:text-base-darker active:text-base-darkest"
        type="button"
      >
        Cancel
      </button>
    </div>
    <div
      class="grid-col-7"
    >
      <div
        class="margin-left-4"
      >
        <button
          class="usa-button usa-button--base"
        >
          Full preview
        </button>
        <h4>
          Preview
        </h4>
        
        <div
          class="overflow-hidden"
        >
          <h2
            class="margin-bottom-1"
          />
          <div
            class="Markdown usa-prose margin-top-0 margin-bottom-4 chartSummaryAbove"
          />
          <div
            class="recharts-responsive-container"
            id=""
            style="width: 100%; height: 300px;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders the VisualizeChart component without horizontal scrolling 1`] = `
<div>
  <div
    class="grid-row width-desktop"
  >
    <div
      class="grid-col-5"
    >
      <div
        class="usa-form-group"
      >
        <label
          class="usa-label text-bold"
          for="title"
        >
          Chart title
        </label>
        <div
          class="usa-hint"
        >
          Give your chart a descriptive title.
        </div>
        <input
          class="usa-input"
          id="title"
          name="title"
          type="text"
          value=""
        />
      </div>
      <div
        class="usa-checkbox"
      >
        <input
          checked=""
          class="usa-checkbox__input"
          id="display-title"
          name="showTitle"
          type="checkbox"
        />
        <label
          class="usa-checkbox__label"
          for="display-title"
        >
          Show title on dashboard
        </label>
      </div>
      <div
        class="usa-form-group"
      >
        <fieldset
          class="usa-fieldset"
        >
          <legend
            class="usa-sr-only"
          >
            Chart type
          </legend>
          <span
            class="usa-label text-bold"
          >
            Chart type
          </span>
          <div
            class="usa-hint"
          >
            Choose a chart type.
          </div>
          <div
            class="margin-top-2"
          >
            <div
              class="usa-radio"
            >
              <input
                class="usa-radio__input"
                id="BarChart"
                name="chartType"
                type="radio"
                value="BarChart"
              />
              <label
                class="usa-radio__label"
                for="BarChart"
              >
                Bar
              </label>
            </div>
            <div
              class="usa-radio"
            >
              <input
                class="usa-radio__input"
                id="ColumnChart"
                name="chartType"
                type="radio"
                value="ColumnChart"
              />
              <label
                class="usa-radio__label"
                for="ColumnChart"
              >
                Column
              </label>
            </div>
            <div
              class="usa-radio"
            >
              <input
                checked=""
                class="usa-radio__input"
                id="LineChart"
                name="chartType"
                type="radio"
                value="LineChart"
              />
              <label
                class="usa-radio__label"
                for="LineChart"
              >
                Line
              </label>
            </div>
            <div
              class="usa-radio"
            >
              <input
                class="usa-radio__input"
                id="PartWholeChart"
                name="chartType"
                type="radio"
                value="PartWholeChart"
              />
              <label
                class="usa-radio__label"
                for="PartWholeChart"
              >
                Part-to-whole
              </label>
            </div>
          </div>
        </fieldset>
      </div>
      <div
        class="margin-top-3 grid-col-6"
      >
        <div
          class="usa-form-group"
        >
          <label
            class="usa-label text-bold"
            for="sortData"
          >
            Sort data
          </label>
          <div
            class="usa-hint"
            id="event-date-start-hint"
          />
          <select
            class="usa-select"
            id="sortData"
            name="sortData"
          >
            <option
              selected=""
              value=""
            >
              Select an option
            </option>
            <option
              value="id###asc"
            >
              "id" A-Z
            </option>
            <option
              value="id###desc"
            >
              "id" Z-A
            </option>
            <option
              value="name###asc"
            >
              "name" A-Z
            </option>
            <option
              value="name###desc"
            >
              "name" Z-A
            </option>
            <option
              value="updatedAt###asc"
            >
              "updatedAt" A-Z
            </option>
            <option
              value="updatedAt###desc"
            >
              "updatedAt" Z-A
            </option>
          </select>
        </div>
      </div>
      <div
        hidden=""
      >
        <legend
          class="usa-legend text-bold"
        >
          Line chart options
        </legend>
        <div
          class="usa-checkbox"
        >
          <input
            checked=""
            class="usa-checkbox__input"
            id="horizontalScroll"
            name="horizontalScroll"
            type="checkbox"
          />
          <label
            class="usa-checkbox__label"
            for="horizontalScroll"
          >
            Display with horizontal scroll
          </label>
        </div>
      </div>
      <div
        class="usa-form-group"
      >
        <label
          class="usa-label text-bold"
          for="summary"
        >
          Chart summary - optional
        </label>
        <div
          class="usa-hint"
        >
          Give your chart a summary to explain it in more depth. It can also be read by screen readers to describe the chart for those with visual impairments. This field supports markdown.
           
          <a
            aria-label=""
            class="link"
            href="/admin/markdown"
            target="_blank"
          >
            <span>
              View Markdown Syntax
            </span>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-external-link-alt fa-w-16 fa-xs margin-left-05"
              data-icon="external-link-alt"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                fill="currentColor"
              />
            </svg>
          </a>
        </div>
        <textarea
          class="usa-textarea"
          id="summary"
          name="summary"
          rows="5"
          style="height: auto;"
        />
      </div>
      <div
        class="usa-checkbox"
      >
        <input
          class="usa-checkbox__input"
          id="summary-below"
          name="summaryBelow"
          type="checkbox"
        />
        <label
          class="usa-checkbox__label"
          for="summary-below"
        >
          Show summary below chart
        </label>
      </div>
      <br />
      <br />
      <hr />
      <button
        class="usa-button usa-button--outline"
        type="button"
      >
        Back
      </button>
      <button
        class="usa-button usa-button--base"
        disabled=""
        type="submit"
      >
        Add Chart
      </button>
      <button
        class="usa-button usa-button--unstyled text-base-dark hover:text-base-darker active:text-base-darkest"
        type="button"
      >
        Cancel
      </button>
    </div>
    <div
      class="grid-col-7"
    >
      <div
        class="margin-left-4"
      >
        <button
          class="usa-button usa-button--base"
        >
          Full preview
        </button>
        <h4>
          Preview
        </h4>
        
        <div
          class="overflow-hidden"
        >
          <h2
            class="margin-bottom-1"
          />
          <div
            class="Markdown usa-prose margin-top-0 margin-bottom-4 chartSummaryAbove"
          />
          <div
            class="recharts-responsive-container"
            id=""
            style="width: 100%; height: 300px;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
